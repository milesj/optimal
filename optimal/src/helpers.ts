import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
