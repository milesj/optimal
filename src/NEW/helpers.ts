import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
