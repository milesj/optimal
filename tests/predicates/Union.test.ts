import {
  union,
  custom,
  array,
  bool,
  object,
  instance,
  number,
  shape,
  string,
  UnionPredicate,
} from '../../src';
import { runChecks, runInProd } from '../helpers';

describe('UnionPredicate', () => {
  let predicate: UnionPredicate;

  class Foo {}
  class Bar {}

  beforeEach(() => {
    predicate = union(
      [
        array(string()),
        bool(true).only(),
        instance(Foo),
        number().between(0, 5),
        object(number()),
        string('foo').oneOf(['foo', 'bar', 'baz']),
      ],
      '',
    );
  });

  it('errors if a non-array is not passed', () => {
    expect(() => {
      // @ts-ignore
      union('foo', []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an empty array is passed', () => {
    expect(() => {
      union([], []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an array with non-predicates is passed', () => {
    expect(() => {
      // @ts-ignore
      union([123], []);
    }).toThrowErrorMatchingSnapshot();
  });

  it('doesnt error if a predicate array is passed', () => {
    expect(() => {
      union([string()], []);
    }).not.toThrow();
  });

  it('sets default value', () => {
    expect(union([string()], 'bar').defaultValue).toBe('bar');
  });

  describe('run()', () => {
    it('errors if a unsupported type is used', () => {
      expect(() => {
        runChecks(union([string(), number(), bool()], []), []);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a nested union is used', () => {
      expect(() => {
        runChecks(
          union<string[]>(
            [string('foo').oneOf(['foo', 'bar', 'baz']), union([number(), bool()], [])],
            [],
          ),
          [],
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors with the class name for instance checks', () => {
      expect(() => {
        runChecks(union([number(), instance(Buffer)], {}), {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns default value if value is undefined', () => {
      predicate.defaultValue = 1;

      expect(runChecks(predicate)).toEqual(1);
    });

    it('returns default value from factory if value is undefined', () => {
      // @ts-ignore
      predicate.defaultValueFactory = () => 'foo';

      expect(runChecks(predicate)).toEqual('foo');
    });

    it('runs array check', () => {
      expect(() => {
        runChecks(predicate, [123]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs boolean check', () => {
      expect(() => {
        runChecks(predicate, false);
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs custom check', () => {
      expect(() => {
        runChecks(
          union(
            [
              string(),
              custom(value => {
                if (typeof value === 'number') {
                  throw new TypeError('Encountered a number!');
                }
              }, ''),
            ],
            0,
          ),
          123,
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs instance check', () => {
      expect(() => {
        runChecks(predicate, new Bar());
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs number check', () => {
      expect(() => {
        runChecks(predicate, 10);
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs object check', () => {
      expect(() => {
        runChecks(predicate, { foo: 'foo' });
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs shape check', () => {
      expect(() => {
        runChecks(
          union(
            [
              shape({
                foo: string(),
                bar: number(),
              }),
            ],
            {},
          ),
          { foo: 123 },
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs string check', () => {
      expect(() => {
        runChecks(predicate, 'qux');
      }).toThrowErrorMatchingSnapshot();
    });

    it('runs correctly for valid values', () => {
      expect(runChecks(predicate, 'foo')).toBe('foo');
      expect(runChecks(predicate, 3)).toBe(3);
      expect(runChecks(predicate, true)).toBe(true);
    });

    it('supports multiple array predicates', () => {
      predicate = union([array(string()), array(number())], []);

      expect(() => {
        runChecks(predicate, [true]);
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(predicate, [123]);
      }).not.toThrow();

      expect(() => {
        runChecks(predicate, ['abc']);
      }).not.toThrow();
    });

    it('supports multiple object predicates', () => {
      predicate = union([object(string()), object(number())], {});

      expect(() => {
        runChecks(predicate, { foo: true });
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(predicate, { foo: 123 });
      }).not.toThrow();

      expect(() => {
        runChecks(predicate, { foo: 'abc' });
      }).not.toThrow();
    });

    it('supports object and shape predicates in parallel', () => {
      predicate = union(
        [
          shape({
            foo: string(),
            bar: number(),
            baz: bool(),
          }).exact(),
          object(string()),
        ],
        {},
      );

      expect(() => {
        runChecks(predicate, { unknown: true });
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(predicate, { foo: 123 });
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(predicate, { foo: 'abc', bar: 'abc', baz: 123 });
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        runChecks(predicate, { foo: 'abc', bar: 123 });
      }).not.toThrow();

      expect(() => {
        runChecks(predicate, { key: 'value' });
      }).not.toThrow();
    });

    it('returns shapes as their full objects', () => {
      predicate = union(
        [
          shape({
            foo: string().required(),
            bar: number(),
            baz: bool(),
          }).exact(),
          object(number()),
        ],
        {},
      );

      expect(runChecks(predicate, {})).toEqual({});
      expect(runChecks(predicate, { foo: 'foo' })).toEqual({
        foo: 'foo',
        bar: 0,
        baz: false,
      });
      expect(runChecks(predicate, { a: 1, b: 2 })).toEqual({ a: 1, b: 2 });
    });

    it('returns an array of shapes as their full objects', () => {
      predicate = union(
        [
          array(
            shape({
              foo: string(),
              bar: number(),
              baz: bool(),
            }).exact(),
          ),
        ],
        [],
      );

      expect(runChecks(predicate, [])).toEqual([]);
      expect(runChecks(predicate, [{ foo: 'foo' }, { bar: 123 }, { baz: true }])).toEqual([
        {
          foo: 'foo',
          bar: 0,
          baz: false,
        },
        {
          foo: '',
          bar: 123,
          baz: false,
        },
        {
          foo: '',
          bar: 0,
          baz: true,
        },
      ]);
    });

    describe('production', () => {
      it(
        'returns default value if value is undefined',
        runInProd(() => {
          expect(runChecks(predicate)).toBe('');
        }),
      );

      it(
        'bypasses checks and returns value',
        runInProd(() => {
          expect(
            runChecks(
              predicate,
              // @ts-ignore Test invalid type
              'qux',
            ),
          ).toBe('qux');
        }),
      );
    });
  });

  describe('typeAlias()', () => {
    it('returns all the available type aliases separated by pipes', () => {
      expect(union([string(), number(), bool()], []).typeAlias()).toBe('string | number | boolean');
    });

    it('supports complex structures', () => {
      expect(predicate.typeAlias()).toBe(
        'array<string> | boolean | Foo | number | object<number> | string',
      );
    });
  });
});
