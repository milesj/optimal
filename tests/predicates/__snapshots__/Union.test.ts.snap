// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnionPredicate errors if a non-array is not passed 1`] = `"A non-empty array of blueprints are required for a union."`;

exports[`UnionPredicate errors if an array with non-predicates is passed 1`] = `"A non-empty array of blueprints are required for a union."`;

exports[`UnionPredicate errors if an empty array is passed 1`] = `"A non-empty array of blueprints are required for a union."`;

exports[`UnionPredicate run() errors if a nested union is used 1`] = `"Invalid field \\"key\\". Nested unions are not supported."`;

exports[`UnionPredicate run() errors if a unsupported type is used 1`] = `"Invalid field \\"key\\". Type must be one of: string, number, boolean."`;

exports[`UnionPredicate run() errors with the class name for instance checks 1`] = `"Invalid field \\"key\\". Type must be one of: number, Buffer."`;

exports[`UnionPredicate run() runs array check 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, boolean, Foo, number, object<number>, string. Received array with the following invalidations:
 - Invalid field \\"key[0]\\". Must be a string."
`;

exports[`UnionPredicate run() runs boolean check 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, boolean, Foo, number, object<number>, string. Received boolean with the following invalidations:
 - Value may only be \\"true\\"."
`;

exports[`UnionPredicate run() runs custom check 1`] = `
"Invalid field \\"key\\". Type must be one of: string, custom. Received number with the following invalidations:
 - Encountered a number!"
`;

exports[`UnionPredicate run() runs instance check 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, boolean, Foo, number, object<number>, string. Received instance with the following invalidations:
 - Must be an instance of \\"Foo\\"."
`;

exports[`UnionPredicate run() runs number check 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, boolean, Foo, number, object<number>, string. Received number with the following invalidations:
 - Number must be between 0 and 5."
`;

exports[`UnionPredicate run() runs object check 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, boolean, Foo, number, object<number>, string. Received object with the following invalidations:
 - Invalid field \\"key.foo\\". Must be a number."
`;

exports[`UnionPredicate run() runs shape check 1`] = `
"Invalid field \\"key\\". Type must be one of: shape. Received object with the following invalidations:
 - Invalid field \\"key.foo\\". Must be a string."
`;

exports[`UnionPredicate run() runs string check 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, boolean, Foo, number, object<number>, string. Received string with the following invalidations:
 - String must be one of: foo, bar, baz"
`;

exports[`UnionPredicate run() supports multiple array predicates 1`] = `
"Invalid field \\"key\\". Type must be one of: array<string>, array<number>. Received array with the following invalidations:
 - Invalid field \\"key[0]\\". Must be a string.
 - Invalid field \\"key[0]\\". Must be a number."
`;

exports[`UnionPredicate run() supports multiple object predicates 1`] = `
"Invalid field \\"key\\". Type must be one of: object<string>, object<number>. Received object with the following invalidations:
 - Invalid field \\"key.foo\\". Must be a string.
 - Invalid field \\"key.foo\\". Must be a number."
`;

exports[`UnionPredicate run() supports object and shape predicates in parallel 1`] = `
"Invalid field \\"key\\". Type must be one of: shape, object<string>. Received object with the following invalidations:
 - Unknown \\"key\\" fields: unknown.
 - Invalid field \\"key.unknown\\". Must be a string."
`;

exports[`UnionPredicate run() supports object and shape predicates in parallel 2`] = `
"Invalid field \\"key\\". Type must be one of: shape, object<string>. Received object with the following invalidations:
 - Invalid field \\"key.foo\\". Must be a string."
`;

exports[`UnionPredicate run() supports object and shape predicates in parallel 3`] = `
"Invalid field \\"key\\". Type must be one of: shape, object<string>. Received object with the following invalidations:
 - Invalid field \\"key.bar\\". Must be a number.
 - Invalid field \\"key.baz\\". Must be a string."
`;
